File: js-sdk/src/tracker.js
import FyipeListener from './listener';
import Util from './util';
import { v4 as uuidv4 } from 'uuid';
import axios from 'axios';
import { name, version } from '../package.json';

class ErrorTracker {
    // constructor to set up global listeners
    #utilObj;
    #listenerObj;
    #eventId;
    #tags = [];
    #extras = [];
    #isWindow = false;
    #fingerprint = [];
    #options = {
        maxTimeline: 5,
        captureCodeSnippet: true,
    };
    #MAX_ITEMS_ALLOWED_IN_STACK = 100;
    #configKeys = ['baseUrl'];
    #event;
    #errorTrackerId;
    #errorTrackerKey;
    #apiUrl;

    constructor(apiUrl, errorTrackerId, errorTrackerKey, options = {}) {
        this._setErrorTrackerId(errorTrackerId);
        this._setApiUrl(apiUrl);
        this._setErrorTrackerKey(errorTrackerKey);
        // set up option
        this._setUpOptions(options);
        this._setEventId();
        this.#isWindow = typeof window !== 'undefined';
        this.#listenerObj = new FyipeListener(
            this.getEventId(),
            this.#isWindow,
            this.#options
        ); // Initialize Listener for timeline
        this.#utilObj = new Util(this.#options);
        // set up error listener
        if (this.#isWindow) {
            this._setUpErrorListener();
        } else {
            this._setUpNodeErrorListener();
        }
    }
    _setErrorTrackerId(errorTrackerId) {
        this.#errorTrackerId = errorTrackerId;
    }
    _setErrorTrackerKey(errorTrackerKey) {
        this.#errorTrackerKey = errorTrackerKey;
    }
    _setApiUrl(apiUrl) {
        this.#apiUrl = `${apiUrl}/error-tracker/${this.#errorTrackerId}/track`;
    }
    _setUpOptions(options) {
        for (const [key, value] of Object.entries(options)) {
            // proceed with current key if it is not in the config keys
            if (!this.#configKeys.includes(key)) {
                // if key is in allowed options keys
                if (this.#options[key]) {
                    // set max timeline properly after checking conditions
                    if (key === 'maxTimeline') {
                        const allowedValue =
                            value > this.#MAX_ITEMS_ALLOWED_IN_STACK ||
                            value < 1
                                ? this.#MAX_ITEMS_ALLOWED_IN_STACK
                                : value;

                        this.#options[key] = allowedValue;
                    } else if (key === 'captureCodeSnippet') {
                        const isBoolean = typeof value === 'boolean'; // check if the passed value is a boolean
                        // set boolean value if boolean or set default `true` if annything other than boolean is passed
                        this.#options[key] = isBoolean ? value : true;
                    } else {
                        this.#options[key] = value;
                    }
                }
            }
        }
    }
    _setEventId() {
        this.#eventId = uuidv4();
    }
    getEventId() {
        return this.#eventId;
    }
    setTag(key, value) {
        if (!(typeof key === 'string') || !(typeof value === 'string')) {
            return 'Invalid Tags type';
        }
        // get the index if the key exist already
        const index = this.#tags.findIndex(tag => tag.key === key);
        if (index !== -1) {
            // replace value if it exist
            this.#tags[index].value = value;
        } else {
            // push key and value if it doesnt
            this.#tags = [...this.#tags, { key, value }];
        }
    }
    // pass an array of tags
    setTags(tags) {
        if (!Array.isArray(tags)) {
            return 'Invalid Tags type';
        }
        tags.forEach(element => {
            if (element.key && element.value) {
                this.setTag(element.key, element.value);
            }
        });
    }
    _getTags() {
        return this.#tags;
    }
    setExtras(extras) {
        extras.forEach(element => {
            if (element.key && element.extra) {
                this.setExtra(element.key, element.extra);
            }
        });
    }

    setExtra(key, extra) {
        this.#extras = { ...this.#extras, [key]: extra };
    }
    setFingerprint(keys) {
        if (!(typeof keys === 'string') && !Array.isArray(keys)) {
            return 'Invalid Fingerprint Format';
        }
        this.#fingerprint = keys ? (Array.isArray(keys) ? keys : [keys]) : [];
    }
    _getFingerprint(errorMessage) {
        // if no fingerprint exist currently
        if (this.#fingerprint.length < 1) {
            // set up finger print based on error since none exist
            this.setFingerprint(errorMessage);
        }
        return this.#fingerprint;
    }
    // set up error listener
    _setUpErrorListener() {
        const _this = this;
        window.onerror = async function(message, file, line, col, error) {
            const errorEvent = { message, file, line, col, error };

            const string = errorEvent.message
                ? errorEvent.message.toLowerCase()
                : errorEvent.toLowerCase();
            const substring = 'script error';
            if (string.indexOf(substring) > -1) {
                return; // third party error
            } else {
                // construct the error object
                const errorObj = await _this.#utilObj._getErrorStackTrace(
                    errorEvent
                );

                // set the a handled tag
                _this.setTag('handled', 'false');
                // prepare to send to server
                _this.prepareErrorObject('error', errorObj);

                // send to the server
                _this.sendErrorEventToServer();
            }
        };
    }
    _setUpNodeErrorListener() {
        const _this = this;
        process
            .on('uncaughtException', err => {
                // display for the user
                // eslint-disable-next-line no-console
                console.log(`${err}`);
                // any uncaught error
                _this._manageErrorNode(err);
            })
            .on('unhandledRejection', err => {
                // display this for the user
                // eslint-disable-next-line no-console
                console.log(`UnhandledPromiseRejectionWarning: ${err.stack}`);
                // any unhandled promise error
                _this._manageErrorNode(err);
            });
    }
    async _manageErrorNode(error) {
        // construct the error object
        const errorObj = await this.#utilObj._getErrorStackTrace(error);

        // set the a handled tag
        this.setTag('handled', 'false');
        // prepare to send to server
        this.prepareErrorObject('error', errorObj);

        // send to the server
        return this.sendErrorEventToServer();
    }
    addToTimeline(category, content, type) {
        const timeline = {
            category,
            data: {
                content,
            },
            type,
        };
        this.#listenerObj.logCustomTimelineEvent(timeline);
    }
    getTimeline() {
        return this.#listenerObj.getTimeline();
    }
    captureMessage(message) {
        // set the a handled tag
        this.setTag('handled', 'true');
        this.prepareErrorObject('message', { message });

        // send to the server
        return this.sendErrorEventToServer();
    }
    async captureException(error) {
        // construct the error object
        const errorObj = await this.#utilObj._getErrorStackTrace(error);

        // set the a handled tag
        this.setTag('handled', 'true');

        this.prepareErrorObject('exception', errorObj);

        // send to the server
        return this.sendErrorEventToServer();
    }
    _setHost() {
        if (this.#isWindow) {
            // Web apps
            this.setTag('url', window.location.origin);
        } else {
            // JS Backend
            // TODO create a way to get host on the backend
        }
    }
    prepareErrorObject(type, errorStackTrace) {
        // log event
        const content = {
            message: errorStackTrace.message,
        };
        this.#listenerObj.logErrorEvent(content, type);
        // set the host as a tag to be used later
        this._setHost();
        // get current timeline
        const timeline = this.getTimeline();
        // get device location and details
        const deviceDetails = this.#utilObj._getUserDeviceDetails();
        const tags = this._getTags();
        const fingerprint = this._getFingerprint(errorStackTrace.message); // default fingerprint will be the message from the error stacktrace
        // get event ID
        // Temporary display the state of the error stack, timeline and device details when an error occur
        // prepare the event so it can be sent to the server
        this.#event = {
            type,
            timeline,
            exception: errorStackTrace,
            deviceDetails,
            eventId: this.getEventId(),
            tags,
            fingerprint,
            errorTrackerKey: this.#errorTrackerKey,
            sdk: this.getSDKDetails(),
        };
    }
    async sendErrorEventToServer() {
        let content;
        await this._makeApiRequest(this.#event)
            .then(response => {
                content = response;
                // generate a new event Id
                this._setEventId();
                // clear the timeline after a successful call to the server
                this._clear(this.getEventId());
            })
            .catch(error => (content = error));
        return content;
    }
    _makeApiRequest(data) {
        return new Promise((resolve, reject) => {
            axios
                .post(this.#apiUrl, data)
                .then(res => {
                    resolve(res);
                })
                .catch(err => {
                    reject(err);
                });
        });
    }
    getCurrentEvent() {
        return this.#event;
    }
    getSDKDetails() {
        return { name, version };
    }
    _clear(newEventId) {
        // clear tags
        this.#tags = [];
        // clear extras
        this.#extras = [];
        // clear fingerprint
        this.#fingerprint = [];
        // clear timeline
        this.#listenerObj.clearTimeline(newEventId);
    }
}
export default ErrorTracker;

File: python-sdk/fyipe_sdk/fyipe_sdk/tracker.py

from .util import Util
from .fyipeListener import FyipeListener
from .fyipeTransport import FyipeTransport
import sys
import logging
import traceback

class FyipeTracker:
    def __init__(self, apiUrl, errorTrackerId, errorTrackerKey, options = {}):
        self.configKeys = ['baseUrl'];
        self.MAX_ITEMS_ALLOWED_IN_STACK = 100
        self.options = {
            'maxTimeline': 5,
            'captureCodeSnippet': True
        }
        self.errorTrackerId = errorTrackerId
        self.errorTrackerKey = errorTrackerKey
        self.apiUrl = apiUrl + "/error-tracker/" + errorTrackerId + "/track"
        self.apiTransport = FyipeTransport(self.apiUrl)
        self.tags = []
        self.fingerprint = []
        self.setUpOptions(options)
        self.util = Util(self.options)
        self.setEventId()
        self.listenerObj = FyipeListener(self.eventId, self.options)

        # initialize exception handler listener
        sys.excepthook = self.custom_excepthook

    # set up options     
    def setUpOptions(self, options):
        """
        Set up options needed for Fyipe Tracker
        """
        if(isinstance(options, dict) != True):
            return # ignore passed options if it is not an object
        
        for option in options:
            value = options[option]
            # proceed with current key if it is not in the config keys
            if option not in self.configKeys:
                # if key is allowed in options
                if self.options[option] is not None:
                    # set max timeline properly after hecking conditions
                    if option == 'maxTimeline':
                        allowedValue = value
                        if value > self.MAX_ITEMS_ALLOWED_IN_STACK or value < 1 :
                            allowedValue = self.MAX_ITEMS_ALLOWED_IN_STACK
                        
                        self.options[option] = allowedValue
                    if option == 'captureCodeSnippet':
                        defaultVal = True
                        # set boolean value if boolean or set default `true` if annything other than boolean is passed
                        if isinstance(value, bool):
                            defaultVal = value
                        self.options[option] = defaultVal
                    else:
                        self.options[option] = value
                            



    def setEventId(self):
        self.eventId = self.util.v4()
    
    def getEventId(self):
        return self.eventId
    
    def addToTimeline(self, category, content, type):
        timelineObj =  {
            "category": category,
            "data": content,
            "type": type
        }

        self.listenerObj.logCustomTimelineEvent(timelineObj);

    def getTimeline(self):
        return self.listenerObj.getTimeline();
    
    def setTag(self, key, value):
        if ( not (isinstance(key, str) or isinstance(value, str)) ):
            raise Exception("Invalid Tag")
        
        exist = False
        for tag in self.tags:
            if(tag['key'] == key):
                # set the round flag
                exist = True
                # replace value if it exist
                tag["value"] = value
                break
            
        if not exist:
            # push key and value if it doesnt
            tag = {
                "key": key,
                "value": value
            }
            self.tags.append(tag)
        
    def setTags(self, tags):
    
        if (not isinstance(tags, list) ):
            raise Exception("Invalid Tags")
        
        for tag in tags:
            if(tag['key'] is not None and tag['value'] is not None):
                self.setTag(tag['key'], tag['value'])

    def getTags(self):
        return self.tags
    
    def setFingerPrint(self, key):
    
        if not isinstance(key, (str, list)):
            raise Exception("Invalid Fingerprint")

        fingerprint = key
        if isinstance(key, str):
            fingerprint = [key]
        
        self.fingerprint = fingerprint
    
    def getFingerprint(self, errorMessage):
    
        # if no fingerprint exist currently
        if (len(self.fingerprint) < 1):
            # set up finger print based on error since none exist
            self.setFingerPrint(errorMessage)
        
        return self.fingerprint
    
    def custom_excepthook(self, exc_type, exc_value, exc_traceback):
        # Do not print exception when user cancels the program
        if issubclass(exc_type, KeyboardInterrupt):
            sys.__excepthook__(exc_type, exc_value, exc_traceback)
            return

        self.captureUncaughtException(exc_type, exc_value, exc_traceback)
        logging.error("An uncaught exception occurred:")
        logging.error("Type: %s", exc_type)
        logging.error("Value: %s", exc_value)

        if exc_traceback:
            format_exception = traceback.format_tb(exc_traceback)
            for line in format_exception:
                logging.error(repr(line))

    def captureMessage(self, message):
        # set the a handled tag
        self.setTag('handled', 'true')
        messageObj = {
            "message": message
        }

        self.prepareErrorObject('message', messageObj);

        # send to the server
        return self.sendErrorEventToServer()
    
    def prepareErrorObject(self, eventType, errorStackTrace):   
        # set a last timeline as the error message
        self.listenerObj.logErrorEvent(errorStackTrace["message"], eventType)
        
        # get current timeline
        timeline = self.getTimeline()
        # TODO get device location and details
        
        tags = self.getTags()
        fingerprint = self.getFingerprint(errorStackTrace["message"]) # default fingerprint will be the message from the error stacktrace
        # get event ID
        # Temporary display the state of the error stack, timeline and device details when an error occur
        # prepare the event so it can be sent to the server
        self.event = {
            "type": eventType,
            "timeline": timeline,
            "exception": errorStackTrace,
            "eventId": self.getEventId(),
            "tags": tags,
            "fingerprint": fingerprint,
            "errorTrackerKey": self.errorTrackerKey,
            "sdk": self.getSDKDetails()

        }
    
    def sendErrorEventToServer(self):
        response = None
        # TODO send to API properly
        response = self.apiTransport.sendErrorEventToServer(self.event)
        # generate a new event Id
        self.setEventId()
        # clear the timeline after a successful call to the server
        self.clear(self.getEventId())
        return response
    

    def getCurrentEvent(self):
        return self.event

    def captureException(self, exception):
    
        # construct the error object
        exceptionObj = self.util.getExceptionStackTrace(exception)

        # set the a handled tag
        self.setTag('handled', 'true')

        self.prepareErrorObject('exception', exceptionObj)

        # send to the server
        return self.sendErrorEventToServer()
    
    def captureUncaughtException(self, exc_type, exc_value, exc_traceback):
        
        # construct the error object
        exceptionObj = self.util.getUncaughtExceptionStackTrace(exc_type, exc_value, exc_traceback)

        # set the a handled tag
        self.setTag('handled', 'false')

        self.prepareErrorObject('exception', exceptionObj)

        # send to the server
        return self.sendErrorEventToServer()
    
    def clear(self, newEventId):
    
        # clear tags
        self.tags = []
        # clear fingerprint
        self.fingerprint = []
        # clear timeline
        self.listenerObj.clearTimeline(newEventId)
    
    def getSDKDetails(self):    
        # default sdk details
        sdkDetail = {
            "name": '',
            "version": ""
        }
        try:
            # ipen set up file to pick values 
            filepath =  'setup.py'

            with open( filepath ) as file:
                for line in file.readlines():
                    if line.startswith("    name"):
                        sdkDetail["name"] = self.__getValueFromLine__(line)
                    if line.startswith("    version"):
                        sdkDetail["version"] = self.__getValueFromLine__(line)
                        

        except Exception as error:
            # something terrible went wrog
            sys.stderr.write( "Warning: Could get SDK version")
        
        return sdkDetail
        
    
    def __getValueFromLine__(self, line):
        # geet the content within the quote 
        start = line.index("\"")
        end = line.rfind("\"")
        return line[start+1: end]








